docs/playbook/ingesting-data/011-ingesting-api-data-2.md

```diff
@@ -0,0 +1,322 @@
+### Desired End Output
+- .Py File
+- Code which can be run frequently. Lambdas have a timeout limit of 15 minutes, so the code must complete within this time. It's therefore likely you'll want your code to run daily.
+- Outputs the **RAW CONTENT** of the API call. No transformations
+- Outputs the content in ` {folder}/import_year={year}/import_month={month}/import_day={day}/import_date={date}/`
+- The dates refer to the date of import, not the date of the Data
+
+### Basic Rules
+
+1. The Python Script needs to be able to determine, on its own, which date range to pull data from
+2. All of the functions need to be contained within the script itself
+3. The Lambda will run the "Lambda_handler" function. This means in place of `__MAIN__()` you will have `def lambda_handler(event, lambda_context):`. Then to run this script, you just have a `lambda_handler("","")` line
+4. If you have variables which you want the Terraform or Glue job to handle, you can pass them through as environment variables instead. `os.environ["TARGET_S3_BUCKET_NAME"] = "landing-zone"`  for example will set the environment variable `TARGET_S3_BUCKET_NAME` to `landing-zone`. `s3_bucket = getenv("TARGET_S3_BUCKET_NAME")` Will then be the code to get the environment variable as a variable in your script
+5. Have no secrets (e.g. API keys) hard coded into the script
+6. Try to use as little external packages as possible, Pandas is around 50mb after you install all of the dependancies. Very expensive
+
+### Some tips
+
+- If you are coding using .py files, the way I have found to code the script is to have 2 separate scripts. One main.py file which has the code to go into the Lambda. Another script which triggers that one and sets environment variables - In Jupyter (So Google Colab). Instead of making multiple scripts, use multiple cells. This is what I would suggest
+	 - Import cell
+	 - Functions cell
+	 - `lambda_handler(event, lambda_context):` which acts as your main(), and interacts with the functions cell
+	 - A cell with lambda_handler("","") and environment variable setters to trigger above. This will not be copied into your .py script when it comes down to making it in the Data Platform but it's what will help trigger your Lambda Handler event
+
+<details><summary>Example Jupyter Script Format</summary>
+<p>
+
+
+```
+import os
+```
+```
+def print_text(text_string):
+	print(text_string)
+```
+```
+def lambda_handler(event, lambda_context):
+	load_dotenv() # load environment variables
+	string_to_print = getenv("STRING_TO_PRINT")
+	print_text(string_to_print)
+
+```
+```
+import os
+os.environ["STRING_TO_PRINT"] = "Bacon" # Variable to be passed via Terraform
+lambda_handler("","")
+```
+
+</p>
+</details>
+
+Once you have your script able to output data from the API locally, we need to modify the script to output to S3.
```

Comment:
I'm not exactly sure about the point you're making here. Are you talking about testing locally while you're developing the script?

Reply:
Yeah, since the script that goes into Lambda is different to a normal script, and it can't really be run normally. I use a 2nd script as a way of settings up all of the env variables as if it was being run in Lambda